share/wake/lib/core/map.wake

```diff
@@ -0,0 +1,548 @@
+# If multiple `Pair`s have the same left value (key), then the resulting `Map`
+# will contain the right value of only the *first* occurrence.
+#
+# *Parameters:*
+#   * `cmpKey`:
+#       A function providing a total ordering over values of the key type.
+#   * `pairs`:
+#       The keys and values which the `Map` should contain.  This does not have
+#       to be sorted, but each left-hand value should be unique.
+#
+# *Examples:*
+#   ```
+#   listToMap scmp ("a" → True, "b" → False, "c" → False, Nil) | msize = 3
+#   listToMap scmp ("a" → 1, "a" → 2, Nil) | mlookup "a" = Some 1
+#   ```
+export def listToMap (cmpKey: k => k => Order) (pairs: List (Pair k v)): Map k v =
+    listToTree (makeCmpPair cmpKey) pairs
+    | Map cmpKey
+
+# Construct a `Map` from the pre-associated key-value pairs in the `Vector`.
+# If multiple `Pair`s have the same left value (key), then the resulting `Map`
+# will contain the right value of only the *first* occurrence.
+#
+# *Parameters:*
+#   * `cmpKey`:
+#       A function providing a total ordering over values of the key type.
+#   * `pairs`:
+#       The keys and values which the `Map` should contain.  This does not have
+#       to be sorted, but each left-hand value should be unique.
+export def vectorToMap (cmpKey: k => k => Order) (pairs: Vector (Pair k v)): Map k v =
+    vectorToTree (makeCmpPair cmpKey) pairs
+    | Map cmpKey
+
+
+# Count how many key-value associations are contained in the `Map`.
+#
+# *Examples:*
+#   ```
+#   mnew scmp | msize = 0
+#   listToMap scmp ("a" → True, "b" → False, "c" → False, Nil) | msize = 3
+#   ```
+export def msize (map: Map k v): Integer =
+    tlen map.getMapData
+
+# Test if the `Map` does not contain any elements.
+#
+# *Examples:*
+#   ```
+#   mnew scmp | mempty = True
+#   listToMap scmp ("a" → True, "b" → False, "c" → False, Nil) | mempty = False
+#   ```
+export def mempty (map: Map k v): Boolean =
+    tempty map.getMapData
+
+
+# Add a given value into the map under the key, if that key does not already exist.
+# Any pair with the same key which already exists in the map *remains unchanged*.
+#
+# For a similar function which uses the new value provided, see `minsertReplace`
+# or `minsertWith`.
+#
+# *Examples:*
+#   ```
+#   mnew scmp | minsert "a" 2 | mlookup "a" = Some 2
+#   listToMap scmp ("a" → 1, Nil) | minsert "a" 2 | mlookup "a" = Some 1
+#   ```
+export def minsert (key: k) (value: v) (map: Map k v): Map k v =
+    editMapData (tinsert (Pair key value)) map
+
+# Add a given value into the map under the key, whether or not it already exists.
+#
+# For a similar function which preserves the original value contained in the
+# map, see `minsert` or `minsertWith`.
+#
+# *Examples:*
+#   ```
+#   mnew scmp | minsertReplace "a" 2 | mlookup "a" = Some 2
+#   listToMap scmp ("a" → 1, Nil) | minsertReplace "a" 2 | mlookup "a" = Some 2
+#   ```
+export def minsertReplace (key: k) (value: v) (map: Map k v): Map k v =
```

Comment:
minor nit, I think replacing is more common so `minsertOnce` might be a better name for the above and this should be `minsert`, alternatively this could also be called `mset`

Reply:
I would definitely agree, but this is the pattern we already had in the `Tree` interface, so I wanted the consistency.